File: dice.rb
#!/usr/bin/env ruby
# dice.rb
# encoding: UTF-8

COLORS = {
  reset: "\e[0m",
  red: "\e[91m",
  green: "\e[92m",
  yellow: "\e[93m",
  blue: "\e[94m",
  magenta: "\e[95m",
  cyan: "\e[96m",
  bold: "\e[1m"
}

def colorize(text, color)
  "#{COLORS[color]}#{text}#{COLORS[:reset]}"
end

DICE_COLORS = [:red, :green, :yellow, :blue, :magenta, :cyan]

def roll_dice(num, faces)
  num.times.map { rand(1..faces) }
end

def format_dice(values, faces)
  values.each_with_index.map do |v, i|
    col = DICE_COLORS[i % DICE_COLORS.length]
    col = :bold if v == 1 || v == faces
    colorize(v.to_s.rjust(2), col)
  end.join(' ')
end

def main
  num_dice = 1
  num_faces = 6
  rolls = 1
  show_sum = false
  show_stats = false
  verbose = false
  output_file = nil

  args = ARGV
  i = 0
  while i < args.size
    arg = args[i]
    case arg
    when '-h', '--help'
      puts "Usage: ruby dice.rb [num_dice] [num_faces] [-r rolls] [-s] [-t] [-v] [-o file]"
      return
    when '-r'
      rolls = args[i+1].to_i if i+1 < args.size
      i += 1
    when '-s'
      show_sum = true
    when '-t'
      show_stats = true
    when '-v'
      verbose = true
    when '-o'
      output_file = args[i+1] if i+1 < args.size
      i += 1
    else
      if num_dice == 1 && arg =~ /^\d+$/ && arg.to_i > 0
        num_dice = arg.to_i
      elsif num_faces == 6 && arg =~ /^\d+$/ && arg.to_i > 1
        num_faces = arg.to_i
      end
    end
    i += 1
  end

  if num_dice < 1 || num_faces < 2
    puts "Количество кубиков >= 1, граней >= 2"
    exit 1
  end

  all_results = rolls.times.map { roll_dice(num_dice, num_faces) }

  lines = []
  if verbose
    all_results.each_with_index do |v, idx|
      sum = v.sum
      dice_str = format_dice(v, num_faces)
      lines << "Бросок #{idx+1}: #{dice_str}  (сумма: #{sum})"
    end
  elsif show_sum
    all_results.each { |v| lines << v.sum.to_s }
  else
    all_results.each { |v| lines << v.join(' ') }
  end

  if show_stats && rolls > 1
    sums = all_results.map(&:sum)
    mn = sums.min
    mx = sums.max
    avg = sums.sum / sums.size.to_f
    sorted = sums.sort
    med = sorted.size.even? ? (sorted[sorted.size/2-1] + sorted[sorted.size/2]) / 2.0 : sorted[sorted.size/2]
    lines << "\nСтатистика по суммам:"
    lines << "  Минимум: #{mn}"
    lines << "  Максимум: #{mx}"
    lines << "  Среднее: #{avg.round(2)}"
    lines << "  Медиана: #{med.round(2)}"
  end

  output = lines.join("\n")
  if output_file
    File.write(output_file, output)
    puts colorize("Результат сохранён в #{output_file}", :green)
  else
    puts output
  end
end

main if __FILE__ == $0

File: dice.js
// dice.js
#!/usr/bin/env node
'use strict';

const fs = require('fs');
const { randomInt } = require('crypto');

const COLORS = {
    reset: '\x1b[0m',
    red: '\x1b[91m',
    green: '\x1b[92m',
    yellow: '\x1b[93m',
    blue: '\x1b[94m',
    magenta: '\x1b[95m',
    cyan: '\x1b[96m',
    bold: '\x1b[1m'
};

function colorize(text, color) {
    return COLORS[color] + text + COLORS.reset;
}

const DICE_COLORS = ['red', 'green', 'yellow', 'blue', 'magenta', 'cyan'];

function rollDice(num, faces) {
    const res = [];
    for (let i = 0; i < num; i++) {
        res.push(randomInt(1, faces + 1));
    }
    return res;
}

function formatDice(values, faces) {
    return values.map((v, i) => {
        let col = DICE_COLORS[i % DICE_COLORS.length];
        if (v === 1 || v === faces) col = 'bold';
        return colorize(String(v).padStart(2), col);
    }).join(' ');
}

function main() {
    const args = process.argv.slice(2);
    let numDice = 1, numFaces = 6, rolls = 1;
    let showSum = false, showStats = false, verbose = false, outputFile = '';

    for (let i = 0; i < args.length; i++) {
        const arg = args[i];
        switch (arg) {
            case '-h':
            case '--help':
                console.log('Usage: node dice.js [num_dice] [num_faces] [-r rolls] [-s] [-t] [-v] [-o file]');
                return;
            case '-r':
                if (i+1 < args.length) rolls = parseInt(args[++i]);
                break;
            case '-s':
                showSum = true;
                break;
            case '-t':
                showStats = true;
                break;
            case '-v':
                verbose = true;
                break;
            case '-o':
                if (i+1 < args.length) outputFile = args[++i];
                break;
            default:
                if (numDice === 1) {
                    const v = parseInt(arg);
                    if (!isNaN(v) && v > 0) numDice = v;
                } else if (numFaces === 6) {
                    const v = parseInt(arg);
                    if (!isNaN(v) && v > 1) numFaces = v;
                }
                break;
        }
    }
    if (numDice < 1 || numFaces < 2) {
        console.log('Количество кубиков >= 1, граней >= 2');
        process.exit(1);
    }

    const allResults = [];
    for (let i = 0; i < rolls; i++) {
        allResults.push(rollDice(numDice, numFaces));
    }

    const lines = [];
    if (verbose) {
        allResults.forEach((v, i) => {
            const sum = v.reduce((a, b) => a + b, 0);
            const diceStr = formatDice(v, numFaces);
            lines.push(`Бросок ${String(i+1).padStart(2)}: ${diceStr}  (сумма: ${sum})`);
        });
    } else if (showSum) {
        allResults.forEach(v => lines.push(String(v.reduce((a, b) => a + b, 0))));
    } else {
        allResults.forEach(v => lines.push(v.join(' ')));
    }

    if (showStats && rolls > 1) {
        const sums = allResults.map(v => v.reduce((a, b) => a + b, 0));
        const mn = Math.min(...sums);
        const mx = Math.max(...sums);
        const avg = sums.reduce((a, b) => a + b, 0) / sums.length;
        const sorted = [...sums].sort((a, b) => a - b);
        let med;
        if (sorted.length % 2 === 0) {
            med = (sorted[sorted.length/2 - 1] + sorted[sorted.length/2]) / 2;
        } else {
            med = sorted[Math.floor(sorted.length/2)];
        }
        lines.push('\nСтатистика по суммам:');
        lines.push(`  Минимум: ${mn}`);
        lines.push(`  Максимум: ${mx}`);
        lines.push(`  Среднее: ${avg.toFixed(2)}`);
        lines.push(`  Медиана: ${med.toFixed(2)}`);
    }

    const output = lines.join('\n');
    if (outputFile) {
        fs.writeFileSync(outputFile, output, 'utf8');
        console.log(colorize(`Результат сохранён в ${outputFile}`, 'green'));
    } else {
        console.log(output);
    }
}

main();
